File: Shared/ConnectionURL.swift
import Foundation
#if canImport(Darwin)
import Darwin
#elseif canImport(Glibc)
import Glibc
#endif

enum ConnectionURL {
    static func parse(_ input: String) -> URL? {
        let trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty,
              trimmed.rangeOfCharacter(from: .whitespacesAndNewlines.union(.controlCharacters)) == nil else { return nil }
        let value = trimmed.contains("://") ? trimmed : "ws://\(trimmed)"
        guard let separator = value.range(of: "://") else { return nil }
        let scheme = value[..<separator.lowerBound].lowercased()
        guard ["ws", "wss", "http", "https"].contains(scheme) else { return nil }
        let rest = value[separator.upperBound...]
        let authority = String(rest.prefix { $0 != "/" && $0 != "?" && $0 != "#" })
        guard let (host, portText) = splitAuthority(authority), validHost(host) else { return nil }
        let secure = scheme == "https" || scheme == "wss"
        let port: Int
        if let portText {
            guard !portText.isEmpty, portText.utf8.allSatisfy({ (48...57).contains($0) }),
                  let explicitPort = Int(portText), (1...65535).contains(explicitPort) else { return nil }
            port = explicitPort
        } else {
            port = secure ? 443 : 8787
        }
        var suffix = String(rest.dropFirst(authority.count))
        guard validSuffix(suffix) else { return nil }
        if suffix.isEmpty || suffix.hasPrefix("?") { suffix = "/" + suffix }
        return URL(string: "\(secure ? "wss" : "ws")://\(host):\(port)\(suffix)")
    }

    static func pairingDestination(from url: URL) -> URL? {
        guard let parts = URLComponents(url: url, resolvingAgainstBaseURL: false),
              parts.scheme?.lowercased() == "stagewand",
              let separator = url.absoluteString.range(of: "://"),
              url.absoluteString[separator.upperBound...].prefix(while: { $0 != "/" && $0 != "?" && $0 != "#" }).lowercased() == "connect",
              parts.percentEncodedPath.isEmpty || parts.percentEncodedPath == "/",
              parts.fragment == nil, let query = parts.percentEncodedQuery,
              query.hasPrefix("url="), !query.contains("&"),
              let destination = String(query.dropFirst(4)).removingPercentEncoding else { return nil }
        return parse(destination)
    }

    private static func splitAuthority(_ authority: String) -> (String, String?)? {
        if authority.hasPrefix("[") {
            guard let end = authority.firstIndex(of: "]") else { return nil }
            let host = String(authority[...end])
            let suffix = authority[authority.index(after: end)...]
            guard suffix.isEmpty || suffix.hasPrefix(":") else { return nil }
            return (host, suffix.isEmpty ? nil : String(suffix.dropFirst()))
        }
        let pieces = authority.split(separator: ":", omittingEmptySubsequences: false)
        guard pieces.count <= 2 else { return nil }
        return (String(pieces[0]), pieces.count == 2 ? String(pieces[1]) : nil)
    }

    private static func validHost(_ host: String) -> Bool {
        if host.hasPrefix("[") {
            guard host.hasSuffix("]") else { return false }
            let address = String(host.dropFirst().dropLast())
            guard !address.contains("%") else { return false }
            if address.contains("."), !validIPv4(String(address.split(separator: ":").last ?? "")) { return false }
            var binary = in6_addr()
            return address.withCString { inet_pton(AF_INET6, $0, &binary) } == 1
        }
        let hostname = host.hasSuffix(".") ? String(host.dropLast()) : host
        guard !hostname.isEmpty, hostname.utf8.count <= 253 else { return false }
        if hostname.contains("."), hostname.utf8.allSatisfy({ (48...57).contains($0) || $0 == 46 }) {
            return validIPv4(hostname)
        }
        return hostname.split(separator: ".", omittingEmptySubsequences: false).allSatisfy { label in
            let bytes = Array(label.utf8)
            return !bytes.isEmpty && bytes.count <= 63 && asciiAlphanumeric(bytes[0]) && asciiAlphanumeric(bytes[bytes.count - 1])
                && bytes.allSatisfy { asciiAlphanumeric($0) || $0 == 45 }
        }
    }

    private static func validIPv4(_ address: String) -> Bool {
        let parts = address.split(separator: ".", omittingEmptySubsequences: false)
        return parts.count == 4 && parts.allSatisfy { part in
            !part.isEmpty && part.count <= 3 && (part.count == 1 || !part.hasPrefix("0"))
                && part.utf8.allSatisfy { (48...57).contains($0) } && (Int(part) ?? 256) <= 255
        }
    }

    private static func asciiAlphanumeric(_ byte: UInt8) -> Bool {
        (48...57).contains(byte) || (65...90).contains(byte) || (97...122).contains(byte)
    }

    private static func validSuffix(_ suffix: String) -> Bool {
        let bytes = Array(suffix.utf8)
        let punctuation = Array("-._~!$&'()*+,;=:@/?".utf8)
        var query = false
        var index = 0
        while index < bytes.count {
            let byte = bytes[index]
            if byte == 63 { query = true }
            if byte == 37 {
                guard index + 2 < bytes.count,
                      bytes[(index + 1)...(index + 2)].allSatisfy({ (48...57).contains($0) || (65...70).contains($0) || (97...102).contains($0) }) else { return false }
                index += 3
                continue
            }
            if byte < 128 && !asciiAlphanumeric(byte) && !punctuation.contains(byte)
                && !(query && (byte == 91 || byte == 93)) { return false }
            index += 1
        }
        return true
    }
}

File: android/protocol/src/main/kotlin/systems/edmundlim/stagewand/protocol/ConnectionURL.kt
package systems.edmundlim.stagewand.protocol

import java.net.URI
import java.net.URISyntaxException
import java.nio.ByteBuffer
import java.nio.charset.CharacterCodingException

object ConnectionURL {
    fun parse(input: String): String? {
        val trimmed = input.trim()
        if (trimmed.isEmpty() || trimmed.any { it.isWhitespace() || it.isISOControl() }) return null
        val value = if ("://" in trimmed) trimmed else "ws://$trimmed"
        val parsed = uri(value) ?: return null
        val scheme = parsed.scheme?.lowercase() ?: return null
        if (scheme !in setOf("ws", "wss", "http", "https") || parsed.rawFragment != null) return null
        val authority = parsed.rawAuthority ?: return null
        val (host, portText) = splitAuthority(authority) ?: return null
        if (!validHost(host)) return null
        val secure = scheme == "https" || scheme == "wss"
        val port = if (portText == null) {
            if (secure) 443 else 8787
        } else {
            if (portText.isEmpty() || portText.any { it !in '0'..'9' }) return null
            portText.toIntOrNull() ?: return null
        }
        if (port !in 1..65535) return null
        val suffix = parsed.toASCIIString().substringAfter("://").drop(authority.length)
            .replace("[", "%5B").replace("]", "%5D")
        val pathAndQuery = if (suffix.isEmpty() || suffix.startsWith("?")) "/$suffix" else suffix
        return "${if (secure) "wss" else "ws"}://$host:$port$pathAndQuery"
    }

    fun pairingDestination(url: String): String? {
        val parsed = uri(url) ?: return null
        if (parsed.scheme?.lowercase() != "stagewand" || parsed.rawAuthority?.lowercase() != "connect") return null
        if (!parsed.rawPath.isNullOrEmpty() && parsed.rawPath != "/") return null
        if (parsed.rawFragment != null) return null
        val query = parsed.rawQuery ?: return null
        if (!query.startsWith("url=") || '&' in query) return null
        val destination = decodeQueryValue(query.drop(4)) ?: return null
        return parse(destination)
    }

    private fun uri(value: String): URI? = try {
        URI(value)
    } catch (_: URISyntaxException) {
        null
    }

    private fun splitAuthority(authority: String): Pair<String, String?>? {
        if (authority.startsWith("[")) {
            val end = authority.indexOf(']')
            if (end < 0) return null
            val suffix = authority.substring(end + 1)
            if (suffix.isNotEmpty() && !suffix.startsWith(":")) return null
            return authority.substring(0, end + 1) to if (suffix.isEmpty()) null else suffix.drop(1)
        }
        val pieces = authority.split(':')
        if (pieces.size > 2) return null
        return pieces[0] to pieces.getOrNull(1)
    }

    private fun validHost(host: String): Boolean {
        if (host.startsWith("[")) {
            if (!host.endsWith("]")) return false
            val address = host.drop(1).dropLast(1)
            if (':' !in address || '%' in address) return false
            if ('.' in address && !validIPv4(address.substringAfterLast(':'))) return false
            return uri("ws://[$address]")?.host != null
        }
        val hostname = host.removeSuffix(".")
        if (hostname.isEmpty() || hostname.length > 253) return false
        if ('.' in hostname && hostname.all { it in '0'..'9' || it == '.' }) return validIPv4(hostname)
        return hostname.split('.').all { label ->
            label.isNotEmpty() && label.length <= 63 && asciiAlphanumeric(label.first()) && asciiAlphanumeric(label.last()) &&
                label.all { asciiAlphanumeric(it) || it == '-' }
        }
    }

    private fun validIPv4(address: String): Boolean {
        val parts = address.split('.')
        return parts.size == 4 && parts.all { part ->
            part.isNotEmpty() && part.length <= 3 && (part.length == 1 || !part.startsWith('0')) &&
                part.all { it in '0'..'9' } && (part.toIntOrNull() ?: 256) <= 255
        }
    }

    private fun asciiAlphanumeric(char: Char): Boolean = char in '0'..'9' || char in 'A'..'Z' || char in 'a'..'z'

    private fun decodeQueryValue(value: String): String? {
        val bytes = java.io.ByteArrayOutputStream()
        var index = 0
        while (index < value.length) {
            if (value[index] == '%') {
                if (index + 2 >= value.length) return null
                val byte = value.substring(index + 1, index + 3).toIntOrNull(16) ?: return null
                bytes.write(byte)
                index += 3
            } else {
                val end = value.indexOf('%', index).let { if (it < 0) value.length else it }
                bytes.write(value.substring(index, end).toByteArray(Charsets.UTF_8))
                index = end
            }
        }
        return try {
            Charsets.UTF_8.newDecoder().decode(ByteBuffer.wrap(bytes.toByteArray())).toString()
        } catch (_: CharacterCodingException) {
            null
        }
    }
}
